opflexagent/opflex_conn_track.py

```diff
@@ -22,16 +24,82 @@
+    using the log level and identity set in a configuration file.
+    The process for this is managed using supervisord, which in
+    version 4.0 or later supports redirecting stdout and stderr
+    to syslog. However, older Linux distributions ship with
+    supervisord versions older than this, so the logger tool
+    must be used for this redirection. While supervisord can
+    handle the lifecycle of the top level process, this class
+    neededs manage the lifecycle of the child prcoesses that
+    are spawned (i.e. conntrack and logger), ensuring that
+    the parent exit code is set correctly so that supervisord
+    can handle any respawning, if needed.
+
+    REVISIT: Should have some sort of pooling to avoid exhaustion
+    """
+    def __init__(self, conntrack_cmd, logger_cmd):
+        self.ret = None
+        self.p1 = self.p2 = None
+        try:
+            # Conntrack process redirects stderr to stdout,
+            # retured Popen has file object for stdout
+            self.p1 = subprocess.Popen(
+                conntrack_cmd, stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT, close_fds=True, shell=True)
+            # Logger uses stdout member from conntrack process for its stdin
+            self.p2 = subprocess.Popen(
+                logger_cmd, stdin=self.p1.stdout, close_fds=True, shell=True)
+        except Exception as e:
+            LOG.error("In running commands: %s: %s" %
+                      (conntrack_cmd + logger_cmd, str(e)))
+
+    def _kill_child_procs(self, save_rc=True):
+        if save_rc:
+            # if poll returns None, it means the process hasn't exited yet
+            # In this case, we're going to kill it, so don't save the RC
+            if self.p1 and self.p1.poll() is not None:
+                # It must have died, so set it to nonzero RC or SIGTERM
+                self.ret = self.p1.returncode or -signal.SIGTERM
+            if not self.ret and self.p2 and self.p2.poll() is not None:
+                # It must have died, so set it to nonzero RC or SIGTERM
+                self.ret = self.p2.returncode or -signal.SIGTERM
+        for proc in [self.p1, self.p2]:
+            LOG.debug("conn_track: proc: %s" % proc.returncode)
+            try:
+                proc.terminate()
+            except Exception:
+                pass
+
+    def terminate(self, signum, frame):
+        if signum is not None:
+            self._kill_child_procs(save_rc=False)
+            sys.exit(0)
+        else:
+            self._kill_child_procs()
+            # what should be the return value here?
+
+    def wait(self):
+        signal.signal(signal.SIGINT, self.terminate)
+        signal.signal(signal.SIGTERM, self.terminate)
+        while (self.p1 and self.p2) and (self.p1.poll() is None and
+                                         self.p2.poll() is None):
+            time.sleep(1)
+
+        self._kill_child_procs()
+        if self.ret is None:
+            self.ret = ''
```

Comment:
will this '' make superviord to restart the conn_track process then?

Reply:
That's the way it was previously, and supervisord handled it.

Reply:
I'm thinking that the above is probably incorrect - from this link (see "exitcodes" section):
http://supervisord.org/configuration.html#program-x-section-settings

supervisord is configured to restart the process unless it gets an exit code of 0 or 2 (https://github.com/noironetworks/python-opflex-agent/blob/master/opflexagent/as_metadata_manager.py#L565). 0 is normal, 2 is SIGINT, which I guess are considered "normal" exit codes. SIGTERM is 15, which is what happens when I kill the the top level process (sys.exit in line 86), and when wait exits due to child process getting killed. I'll change this to drop the empty string.

Reply:
do we still plan to change self.ret = '' to something else then?